First stab at using jade

vendethiel opened the pull request:
I think jade got this place here (unless you dislike it ofc), I just stole your buildfile and added jade for my own crappy userscript so I thought i'd do that. Also includes random coco love.

vendethiel committed:
first pass as using jade

vendethiel committed:
convert ALL the things

vendethiel commented on Cokefile:
```diff
@@ -62,8 +74,11 @@ task \build 'build userscript' ->
+        "(function(){
+        var" #avoid jade creating global
+        read \node_modules/jade/runtime.min.js
+        "jade.escape = function (it) { return it; };"
```
'tis a bit of a hack but we don't need escaping

vendethiel commented on Cokefile:
```diff
@@ -62,8 +74,11 @@ task \build 'build userscript' ->
-        "(function(){"
-        '"use strict";'
+        "(function(){
+        var" #avoid jade creating global
```
'tis a real & ugly hack :(. Had to remove strict for jade's `with` too.

qqueue commented:
The main reason I'm not using a template library any more (I used handlebars in the very early commits from last February or so) is because they were a lot slower than string interpolation.

However, I just fetched your branch, and from my metrics it doesn't look like jade is appreciably slower than what I currently have (maybe because 4chan is no longer quirks mode, which somehow sped up parsing considerably too).

I still want to test some more, but I am definitely willing to pull this if I can be sure it's not that much slower.

vendethiel commented:
nice :). Yes jade is very fast compiled and that's one of the reasons why I love it.
You should probably test it anyway because I only tested it on a board page index - I'm not a 4channer.
I hardly see the benefit of handlebars, that's slow because of the parsing and you don't gain anything for the real cumbersome part (the html).

qqueue commented:
Only tangentially related, but I wrote this small jsperf test just now:

http://jsperf.com/innerhtml-vs-createelement-based-low-level-templating

At one point, I wanted to write a pure DOM-based template compiler, but it looks like it's significantly slower than just writing strings :/

vendethiel commented:
Interestingly enough, dom manips are 2 times faster on my iphone. I'll try to see the numbers for chrome when I'm home.

vendethiel commented:
By the way, Stylus now has a sync API (as per LearnBoost/stylus#934), tell me if you want me to change the way css are compiled.

(actually, the whole code is sync, and you could've done this : 
`var css;
stylus str .render -> css := it`
but it feels like a hack)

vendethiel commented:
Just so you know, I also changed the enhancer to embed youtube videos. Nami-Doc/wowboardhelpers@54388e5.
I also fixed a bug where `(http://)` would get linkified

```
    # linkify links not preceded by a quote or double-quote (should avoid
    # relinkifying href= urls)
    * * /([^"']|^)(https?:\/\/)([^<\s\)]+)/g
```

(I just added `\)` in the exclude pattern)
I also changed youtube regexp to allow for mobile version : 
`(?:(?:www|m)\.)?            # optional subdomain (some people add it)`

qqueue commented:
> sync API

Cool, but I can change it later myself in a different commit.

> enhancer

but you took out all my colorful comments ;_;

Also, you've reminded me that I still haven't properly licensed my code. I'll have to get around to that sometime.

Re: the pull, I did a slightly better test on [this sticky /q/ thread](https://boards.4chan.org/q/res/400834) between my templates and the jade versions. Using my netbook, template + parse (by innerHTML) takes ~1300ms with just strings, and ~1600ms with jade, so while it's definitely _slower_, the question still remains whether developer ease is worth an extra 25% time to interactive on every page load. Confounding factors are the fact that my netbook is pretty slow and I still haven't gotten around to testing jade on my main desktop machine, as well as the sample size of 1. However, ~300ms is pretty significant, especially on every page load.

I bet the slowness is mostly due to the `with` block. It's a shame that jade requires it. Though, it seems like jade's compiler is advanced enough to make generating dot assignments fairly simple, might be an project to hack on.

I really do need to add automated tests and local test data to html5chan soon. Having to go find threads on 4chan that meet certain conditions just to test certain parts of the code is definitely not very good developer feedback. Refactoring html5chan to be testable will definitely take some work though.

vendethiel commented:
300ms is definitely too slow but tbh I wouldn't wait 1.3s for an userscript to begin with.
I removed some comments because they didi't seemed necessary :p.
Jade uses `with` to avoid having to prefix every single js part with `locals`

qqueue commented:
> 1.3s

Such is life on a netbook. Even loading a 300-post thread _without_ extensions takes ~3 seconds.

> Jade uses with to avoid having to prefix every single js part with locals

A big price to pay for minor readability enhancement of _generated_ code. I'll have to take a closer look at jade's compiler this weekend perhaps.

vendethiel commented:
Well 4chan is slow and should cache moar :p. 

> A big price to pay for minor readability enhancement of generated code. I'll have to take a closer look at jade's compiler this weekend perhaps

Not like object access is free either. And such code would prevent using globals in jade templates (not that I care). Well, I have no idea how bad `with` performs. Will hack a jsperf when I can (next class probably :p).
And just so you know, jade parses things you'd expect it no to (`a(title="oh!)")` is a syntax error but does lot parse certain things you'd expect it to (like js exprs).

I still havn't decided on the directory structure, I'm currently using src/:page/{,templates,styles} but I'm not sure for future-proofing and how to handle settings etc. Will have to think about it later.

And (not trying to be a douche but ...) contributing to 4chanX seems to be a fucking pita.

vendethiel commented:
And here's the jsperf.
http://jsperf.com/with-vs-raw-object-access

With 500 it loop :
1,443 ops VS 1,003,319
Without : 
377,511 ops VS 63,400,770
I thought such a simple case would be optimized by caching whether x is `with`d but apparently it's not. Haha.

qqueue commented:
I put better metrics around the initial render process. Here are two nearly identical compiled versions of html5chan, with and without jade:

https://gist.github.com/qqueue/4726278

On the same sticky /q/ thread, string interp does its thing in 52ms, and jade in 322ms (n = 1). Firefox actually parsing and rendering the generated string takes about 450ms in either version.

This is still on my netbook though. I really want to test it on my main machine, but my ubuntu install crapped out last friday, so I'm kinda between OSs at the moment. I'm trying fedora now, but its font rendering is shit compared to ubuntu, among other things (gnome 3). I'm thinking I'll just install debian tomorrow, and hope its stability is actually worth something.

vendethiel commented:
I wanted to say "ubuntu is to unix what vista is to windows" but well.

Firefox seems overly slow. Any reason for that ??

qqueue commented:
> contributing to 4chanX seems to be a fucking pita.

To be fair, changing shit is more risky when you have a userbase larger than one person (._.) I actually enjoy the code review process too. My code is considerably more tight than the initial implementation. 

> such a simple case would be optimized

IIRC, putting `with` statements in function body causes SpiderMonkey/V8 to give up on JIT completely for that function, just like putting in an `eval` call. I remember watching a good presentation on V8's JIT a couple weeks ago, I'll have to find it again. Also of note is that V8 is able to completely optimize for loops even when you don't explicitly cache the length variable, which is cool.

vendethiel commented:
> you have a userbase larger than one person

That's the problem when you have more userscripts than true 4channers :p.

> IIRC, putting with statements in function body causes SpiderMonkey/V8 to give up on JIT completely for that function, just like putting in an eval call. I remember watching a good presentation on V8's JIT a couple weeks ago, I'll have to find it again. Also of note is that V8 is able to completely optimize for loops even when you don't explicitly cache the length variable, which is cool.

To that extent ? Oh well. Might as well check for perfs using `locals.` with the `with` :p.

qqueue commented:
> Firefox seems overly slow.

It's likely just the netbook again.

Also, I broke your pull request with the last few commits. Can you rebase on master?

vendethiel commented:
Uuuuh. I'm getting "mv.exe stopped working errors"
That seems nasty and I'm a bit affraid. Will try again tomorrow ...
Edit : oh god, that broke everything... Will do a new clone an manually merge commits one by one ... Vi/vim isn't even working anymore so I cant squash. I'm fucking gonna throw this piece into the windows...
Weeeell I'll wait for you to decide before doing it.

(My own userscript is used by ~20 people because there's no other wow forum userscript ...
But it's breaking things for ie6/7 compat, I have no API, it's filled with bugs I have to deal with ... Haha)

qqueue commented:
> ie6/7 compat

Ouch.

In the meantime, because I am just so curious, I manually edited the generated jade templates and replaced all the variables with either `locals.` or just the global `board` object. It cut down the runtime by ~50ms, so still not great.

Then, I removed the `escape` function completely, both from the templates and `jade.attrs`. cut down another ~50ms.

Finally, I added a `console.profile()` block around the templating, and it turns out that 50% of the time is spent in the `attrs` function :/ So even if I manage to factor out `with` from jade, I'd also have to rewrite `attrs` to get anywhere close to what I currently have. Thus, thanks for the pull request, but I'm going to have to reject it considering the current performance of jade.

Trying to improve jade's compiler would definitely be a fun project though. You could call it... cut-jade :gem:

Modified scripts:

https://gist.github.com/qqueue/4726278/

qqueue commented:
Finally got around to trying it on my desktop. It's definitely not as bad:
- vanilla html5chan: 21ms generation
- jade-nowith-noescape: 64ms
- jade (plain): 97ms

But still slower by 5% to interactive, unfortunately. And honestly, writing plain HTML for html5chan isn't that bad (at least for now when I've only got 2 views for 2 domain classes...)

But gosh, I am just way too ki ni narimasu for js performance. Gotta go do my homework now.

vendethiel commented:
Hmmm ? Isn't escaping already disabled ?

Well that's _far_ too slow. I'll find a way to allow .html templates in my userscript to start with

qqueue commented:
> escaping already disabled

Even the noop `escape` function was being called so many times that it showed up in the profiler. I just went the extra mile and removed the call to `escape` everywhere.

vendethiel commented:
Worth a flag, like `{-escape}`. Not sure what the attrs function does and if we can add something to inline

vendethiel commented:
Actually, I could've used `!= abc` and `!{interp}` to remove that escape thing.
Might be worth changing `jade.attrs` to something like

``` coffee
jade.attrs = ->
  return unless len = it.length

  buf = []
  for k, v in it
    if k is 'class' and Array.isArray v
      v *= ' '
    buf.push "#k='#v'"
  buf * ' '
```

Gotta refactor my thing to detect .jade or .html in my `templates/`. Will do a PR if I'm satisfied later.

vendethiel commented:
Aaaand it's done.

Nami-Doc/wowboardhelpers@e14c78d3a86119c6a4eb818cc957ba44c0cef55d

It's extensible and my main concern was to extract my templates to outer files.
Tell me if you want me to pr this stuff instead, or feel free to copypasta what you like.

EDIT : actually I might use https://github.com/creationix/haml-js. Or even https://github.com/netzpirat/haml-coffee (might fork that :p)
EDIT² : Done in Nami-Doc/wowboardhelpers@32bbe55e5306607bb39f13fb071cc13d44f06c5e. It was a good fight.
HamlCoffee tries to assign itself the variable so I have to do a dirty replace to add the `return`, option "escapeHtml" is "htmlEscape" in the docs and things but haha, it's ok now.
I'll think about forking + doing a hamlcoco (or hamls :°)

vendethiel commented:
Ping again, because I love pseudo-monologizing.
I edited again your autolink feature because 1) I wanted autolink for images too and 2) I was getting `<a ...><a ...`.
I moved `# specialcase linkify urls without a protocol but with a common tld` juste after `(link` specialcase and everything seems to work fine.

qqueue commented:
> autolink

The ordering requirement is nasty. It's a consequence of using regex on HTML, unfortunately. Other, more robust linkifying scripts use TextNodes and `splitNode` and stuff, but all that DOM bullshit makes my eyes glaze over.

vendethiel commented:
plain horrible and just slow.
I also re-added a colorful comment in it.

Also, if you want me to pr for haml-coffee, tell me. I disabled escaping and it's not using with.
It looks like [this](https://github.com/Nami-Doc/wowboardhelpers/blob/master/wowboardhelpers.user.js#L142) (You can remove the `c$` altogether).

qqueue commented:
[The regex replace calls](https://github.com/Nami-Doc/wowboardhelpers/blob/master/wowboardhelpers.user.js#L147) seem unnecessary, why are those there?

Don't stress out too much about a pull request. The small amount of templating that html5chan currently has is still perfectly manageable with regular string interpolation. If you still want to, I'm going to put my threshold for accepting at 10% overhead over what I currently have (on the "generate HTML" metric). I feel like that's an acceptable cost per page load for increased developer ease.

vendethiel commented:
No idea why you'd clean empty id/class like that. Added a hotfix in my build file.

I don't stress about a PR. That's just a small contribution, and I start thinking having some kind of API etc for these things would be cool.
10% threshold is ok, but just look at one of the generated functions : 

``` js
templates.author = templates['forum-topics/author'] = function(context) {
    return (function() {
      var $c, $o;
      $c = c$;
      $o = [];
      $o.push("<span class='" + ($c(["poster", this.own ? "own-poster" : void 0, this.cm ? "type-blizzard" : void 0])) + "'>");
      $o.push("" + $c(this.name));
      if (this.cm) {
        $o.push("<img src='/wow/static/images/layout/cms/icon_blizzard.gif' alt='' />");
      }
      $o.push("</span>");
      return $o.join("\n").replace(/\s(\w+)='true'/mg, " $1='$1'").replace(/\s(\w+)='false'/mg, '').replace(/\s(?:id|class)=(['"])(\1)/mg, "");
    }).call(context);
  };
```

Here's the part in my slakefile : 

``` coffee
  code = hamlCoffee.template read(it), name, "w.templates", opts
  code = (code.trim! / '\n')[1 to -2] * '\n'

  code #crappy hotfixes :(
    .replace "w.templates['#name'] = " ""
    .replace '$o.join("\\n").replace(/\s(?:id|class)=([\'"])(\1)/mg, "");' '$o.join("")'
    .replace """
var _ref;

  if ((_ref = w.templates) == null) {
    w.templates = {};
  }
""" ""
    .trim!
```

vendethiel commented:
Ah. I just noticed jade can avoid using `with` with the `self:` option.
ie

``` coffee
compile.jade = ->
  jade.compile do
    it.replace('@' 'this.')
    {self: 'this', -compileDebug, +client}
```

Still leaves the `escape` :(. I suppose we have `!{}`.

qqueue commented:
> cleaning empty id/classes

That is curious, isn't it? I guess it's because of this:

``` js
var a = document.createElement('a')
a.id //=> ''
a.className //=> ''
a.classList.length //=> 0
a.outerHTML  //=> '<a></a>'

a.id = ''
a.id //=> ''
a.outerHTML //=> '<a id=""></a>'

a.className = ''
a.className //=> ''
a.classList.length //=> 0
a.outerHTML //=> '<a id="" class=""></a>'
```

In firebug on firefox 18.0.2. Bizarre.

Even more bizarre:

``` js
var a = document.createElement('a')
a.id = ""
a.className = ""

document.body.appendChild(a)

document.getElementById('') //=> null
document.getElementById(a.id) //=> null
document.getElementsByClassName('').length //=> 0
document.getElementsByClassName(a.id).length //=> 0
document.querySelector('[id=""]') == a //=> true
document.querySelector('[class=""]') == a //=> true
```

I wonder if this is spec behavior.

vendethiel commented:
Ah, I think I read something about this in the jquery issue tracker
